### PR TITLE
Add container name to kubernetes env adapter

### PR DESCRIPTION
**Pull request by bianpengyuan**
This will allow kubernetes env adapter to derive container name from destination port if provided.

### CODE REVIEW

File: mixer/adapter/kubernetesenv/template/template.proto [hunk 1]
```diff
@@ -72,6 +72,9 @@ message Template {
     // Destination pod's ip.
     istio.mixer.adapter.model.v1beta1.IPAddress destination_ip = 4;
 
+    // Destination container's port number.
+    int64 destination_port = 7;
```
Comment (douglas-reid) [inline mixer/adapter/kubernetesenv/template/template.proto]: can you add a comment to the proto at the top to show last used field number / next available field number, as these are no longer in order?

Comment (bianpengyuan) [reply mixer/adapter/kubernetesenv/template/template.proto]: Done.

**Comment by bianpengyuan** [comment]
@douglas-reid  Sure, I'd like to add an e2e test in this PR. Should I add it to mixer/test/e2e/e2e_apa_test.go?

**Comment by douglas-reid** [comment]
@bianpengyuan does that test actually involve k8s ?  I was thinking `e2e_mixer`, but am fairly flexible.

**Comment by bianpengyuan** [comment]
@douglas-reid I see. Yeah, adding it to mixer_e2e makes sense. I will add a verification there once proxy reporting port [work](https://github.com/istio/proxy/pull/1327) reaches istio/istio repo.

**Comment by bianpengyuan** [comment]
@douglas-reid Can we merge this PR first? I will add yaml file change and e2e test after this PR to avoid the PR becoming too large.

**Comment by bianpengyuan** [comment]
@douglas-reid Added an e2e test for this change. ptal, thanks!

File: install/kubernetes/helm/istio/charts/mixer/templates/config.yaml [hunk 2]
```diff
@@ -107,8 +107,12 @@ spec:
       valueType: STRING
     source.serviceAccount:
       valueType: STRING
+    source.podName:
```
Comment (douglas-reid) [inline install/kubernetes/helm/istio/charts/mixer/templates/config.yaml]: @geeknoid how should we add these attributes in your proposed new model?

should we be using `source.workload.name` here instead? 

and, below, should these be: `destination.workload.name` and  `destination.workload.containerName`  ?

Do we care that these are k8s-only attributes?

**Comment by geeknoid** [comment]
Could we wait a bit before pushing in this change? Let's discuss the exact attribute names/semantics in https://github.com/istio/istio/issues/4824. Once we're settled there, it'll be easy to update this PR accordingly.

I'm going to add these new attributes to #4824 so we can discuss.

File: install/kubernetes/helm/istio/charts/mixer/templates/config.yaml (hunk 2 above)
Comment (geeknoid) [reply install/kubernetes/helm/istio/charts/mixer/templates/config.yaml]: source.podName ends up being source.name in the updated model.
desrtination.podName is destination.name
destination.containerName is destination.container.name

**Comment by geeknoid** [comment]
@douglas-reid Should we close this PR? I think what you're doing subsumes this, right?

**Comment by douglas-reid** [comment]
@geeknoid we definitely need to align them, but overall taking in port and putting out container info (when available) should be complimentary to the other work.

File: install/kubernetes/helm/istio/charts/mixer/templates/config.yaml [hunk 3]
```diff
@@ -121,6 +125,10 @@ spec:
       valueType: STRING
     destination.serviceAccount:
       valueType: STRING
+    destination.podName:
```
Comment (douglas-reid) [inline install/kubernetes/helm/istio/charts/mixer/templates/config.yaml]: let's delete this and the `source.podName` lines.

File: install/kubernetes/helm/istio/charts/mixer/templates/config.yaml [hunk 4]
```diff
@@ -121,6 +125,10 @@ spec:
...
+      valueType: STRING
+    destination.containerName:
```
Comment (douglas-reid) [inline install/kubernetes/helm/istio/charts/mixer/templates/config.yaml]: this should become `destination.container.name`.

File: install/kubernetes/helm/istio/charts/mixer/templates/config.yaml [hunk 5]
```diff
@@ -429,10 +438,13 @@ spec:
     source.namespace: $out.source_namespace | "default"
     source.service: $out.source_service | "unknown"
     source.serviceAccount: $out.source_service_account_name | "unknown"
+    source.podName: $out.source_pod_name | "unknown"
```
Comment (douglas-reid) [inline install/kubernetes/helm/istio/charts/mixer/templates/config.yaml]: i think if you rebase, you'll see this assignment to `source.name`

**Comment by bianpengyuan** [comment]
@douglas-reid Updated. Ptal.

**Comment by bianpengyuan** [comment]
@douglas-reid the newly added test passed. ptal.

**Pull request merged**